### ISSUE TITLE
My custom rules cannot be skipped using tag "skip_ansible_lint"

### ISSUE THREAD

**Issue by pescobar** [willthames/ansible-lint#307]
# Issue Type
- Bug report

# Ansible and Ansible Lint details

```
ansible 2.4.2.0
ansible-lint 3.4.17
```

- ansible installation method: pip
- ansible-lint installation method: pip

# Desired Behaviour

Being able to use the tag `skip_ansible_lint` for my custom rules

# Actual Behaviour (Bug report only)

I have this custom rule which I use doing `ansible-lint -r`

```
from ansiblelint import AnsibleLintRule
import re


class NoKeyValueFormat(AnsibleLintRule):
    id = 'ANSIBLE0003'
    shortdesc = 'Dont use key=value format for tasks'
    description = 'You should use native YAML syntax instead of key=value format'
    tags = ['formatting']

    def match(self, file, line):
        pattern_name = re.compile('^[\s|-].*name:')
        pattern_when = re.compile('^[\s].*when:')
        pattern_key_value = re.compile('[a-zA-Z0-9_.-]=["{a-zA-Z0-9_.-]')
        if pattern_name.findall(line):
            return False  # if this is the name of the task skip check
        if pattern_when.findall(line):
            return False  # if this is a conditional skip check
        if pattern_key_value.findall(line):
            return True  # search for key=value format in any other lines
```

I have a false positive in one of my tasks and I would like to use `skip_ansible_lint` for this task but it doesn't work.  The custom rule is always applied no matter which tags I apply to the task. 

Is this an error in my custom rule or is it a bug in ansible-lint?

**Comment by willthames** [comment]
you're matching against lines, and skip_ansible_lint works at the task level.

ansible-review has a check for this. It would be possible to convert the check to be used by ansible-lint, I believe, and handle skip_ansible_lint, but I've made no effort to do so

https://github.com/willthames/ansible-review/blob/master/lib/ansiblereview/tasks.py

**Comment by pescobar** [comment]
thanks for the clarification @willthames . I understand now.

Is there any easy way to apply my rule at the task level so I can use `skip_ansible_lint`. Something like this pseudococe?

```
for task in playbook:
    applyMyrule()
```

I think this way I would keep the same functionality and I would be able to use `skip_ansible_lint` but I am not sure if this is easy to do with ansible-lint

**Comment by willthames** [comment]
There are three match functions, one of which matches at the task level.
Have a look at the built-in rules, such as UseBecomeInsteadOfSudo (not at a
computer right now to check)

**Comment by sebthebert** [comment]
@pescobar 

You should use `def matchtask(self, file, task):` instead of `def match(self, file, line):`.

Take a look at the [README Rules section](https://github.com/willthames/ansible-lint#rules).

**Comment by Rylon** [comment]
I did indeed look into using `matchtask`, however my obstacle there was that the task at that point has already been parsed and so rules you wish to apply against the syntax in the playbook can't go there, if I understand correctly?

In my particular case it's a rule to look for any tasks using the older style `key=value`, but inside the `matchtask()` method, `task` is already parsed into a dictionary and so you cannot see what the original form was at that point?

**Comment by willthames** [comment]
@rylon ah, I understand that use case very well - that's #297, which is hard for the reasons that you state.

It's definitely possible, but would require exposing a different kind of rule. I do usually just say "ansible-review does this by design, use that" but I can see the appeal of getting the same check into ansible-lint.

Those are my thoughts, I'll let others decide where it should go on the roadmap.

**Comment by Rylon** [comment]
Thanks :) I'll also take a look at porting over to Ansible Review, do you have any recommendations for approaching that? We're using the built in rules from `ansible-lint` and around 10 or so of our own custom ones, so not a massive amount. If we could match the rules that would be fine I think.

**Comment by willthames** [comment]
@Rylon ansible-review is more of a pick and mix - have a look at the examples - they were heavily based on the rules I used at the time, it should show how to set up your own rules.

**Comment by Rylon** [comment]
Cool, thanks @willthames I'll see if I can take a look this week to migrate over 👍
